subprojects/composite-builds/src/integTest/groovy/org/gradle/integtests/composite/CompositeBuildDependencyCycleIntegrationTest.groovy

```diff
@@ -254,8 +263,202 @@ project(':b1') {
+            includeBuild('../buildC')
+        """
+        buildC.buildFile << """
+            task c {
+                dependsOn gradle.includedBuild('buildB').task(':b')
+            }
+        """
+        buildC.settingsFile << """
+            includeBuild('../buildB')
+        """
+
+        when:
+        resolveFails(":a")
+
+        then:
+        failure.assertHasDescription("""Circular dependency between the following tasks:
+:buildB:b
+\\--- :buildC:c
+     \\--- :buildB:b (*)""")
+    }
+
+    // TODO: this ends up in a deadlock. Figure out if this is a valid use case and what to do with it.
+    @Ignore
+    def "direct mustRunAfter cycle between included builds"() {
+        given:
+        buildA.buildFile << """
+            task a {
+                dependsOn gradle.includedBuild('buildB').task(':b')
+            }
+        """
+        buildB.buildFile << """
+            task b {
+                dependsOn gradle.includedBuild('buildC').task(':c')
+            }
+        """
+        buildB.settingsFile << """
+            includeBuild('../buildC')
+        """
+        buildC.buildFile << """
+            task c {
+                mustRunAfter gradle.includedBuild('buildB').task(':b')
+            }
+        """
+        buildC.settingsFile << """
+            includeBuild('../buildB')
+        """
+
+        when:
+        resolveFails(":a")
+
+        then:
+        failure.assertHasDescription("""Circular dependency between the following tasks:
+:buildB:b
+\\--- :buildC:c
+     \\--- :buildB:b (*)""")
+    }
+
+    // TODO: this ends up in a deadlock. Figure out if this is a valid use case and what to do with it.
+    @Ignore
+    def "indirect mustRunAfter cycle between included builds"() {
+        given:
+        buildA.buildFile << """
+            task a {
+                dependsOn gradle.includedBuild('buildB').task(':b')
+            }
+        """
+        buildB.buildFile << """
+            task b {
+                dependsOn gradle.includedBuild('buildC').task(':c')
+            }
+        """
+        buildB.settingsFile << """
+            includeBuild('../buildC')
+        """
+        buildC.buildFile << """
+            task c {
+                dependsOn gradle.includedBuild('buildD').task(':d')
+            }
+        """
+        buildC.settingsFile << """
+            includeBuild('../buildD')
+        """
+
+        def buildD = singleProjectBuild('buildD')
+        buildD.buildFile << """
+            task d {
+                mustRunAfter gradle.includedBuild('buildB').task(':b')
+            }
+        """
+        buildD.settingsFile << """
+            includeBuild('../buildB')
+        """
+        includedBuilds << buildD
+
+        when:
+        resolveFails(":a")
+
+        then:
+        failure.assertHasDescription("""Circular dependency between the following tasks:
+:buildB:b
+\\--- :buildC:c
+     \\--- :buildD:d
+          \\--- :buildB:b (*)""")
     }
```

Comment:
Those two cases cause a deadlock with 6.7 too - they are not introduced by removing the cycle check.